### DISCUSSION TITLE
When the keys are not known beforehand

### DISCUSSION THREAD

**Discussion by Romi-KS** [wnameless/json-flattener#76] [Q&A] [answered]
I recently came across this great library. but I am still wondering how to use it when only top level and leaf nodes are known for example these are the flatened keys:
previewSquad.India.RohitSharma.score
previewSquad.India.ViratKohli.score
previewSquad.Pakistan.ImranKhan.score
previewSquad.Pakistan.Kadar.score

Country and player_name are changing with every set of JSON.

**Comment by wnameless** [comment, marked as answer]
What is your goal? Do you want to unflatten those data?

**Reply by Romi-KS** [reply]
Thanks for the reply.
No, I am comfortable with flatten/unflatten. :-)
My goal is to get the next key. 
For Examples. 
if my search key is "previewSquad" it should return and array (India,Pakistan). 
if my search is "previewSquad.India" it should return array(RohitSharma,ViratKohli)
Similarly,
for  the keys "previewSquad.Pakistan" it should return array(ImranKhan,Khan)

**Reply by wnameless** [reply]
First, I assume that you have a complete JSON input. Here is the solution in Java 17:
```java
    String json = """
        {
          "previewSquad.India.RohitSharma.score": 0,
          "previewSquad.India.ViratKohli.score": 0,
          "previewSquad.Pakistan.ImranKhan.score": 0,
          "previewSquad.Pakistan.Kadar.score": 0
        }
        """;

    var map = JsonUnflattener.unflattenAsMap(json);
    var subMap = map.get("previewSquad");
    if (subMap instanceof Map m) {
      System.out.println(m.keySet());  // [India, Pakistan]
    }
```

More generic approach:
```java
  public List<String> findKeys(Map<String, ?> map, String search) {
    var keys = new ArrayList<String>();
    Collections.addAll(keys, search.split(Pattern.quote(".")));

    Map<?, ?> subMap = map;
    while (!keys.isEmpty()) {
      var key = keys.get(0);
      if (subMap.get(key) instanceof Map<?, ?> m) {
        subMap = m;
        keys.remove(0);
      } else {
        break;
      }
    }

    if (subMap != map && keys.isEmpty()) {
      return subMap.keySet().stream().map(Object::toString).toList();
    }
    return Collections.emptyList();
  }

  System.out.println(findKeys(map, "previewSquad.India")); // [RohitSharma, ViratKohli]
  System.out.println(findKeys(map, "previewSquad.Pakistan.Kadar")); // [score]
```
